### PR TITLE
IRepository documentation

### LINKED ISSUE #574: Update repository system - Explore FileRepository
## Parent Task: 
- https://github.com/f1nal3/Juniorgram/issues/401 

## Description:
Currently, we have a repository system, that perform unified interface for work with DB. One of the implementations is _FileRepository_.
Explore interface of _repository system_ and make a list of problems for finishing this module.

## Acceptance Criteria:
Readiness is determined by @Orange-hanter  after the presentation at the meeting.

### PR DESCRIPTION
**Pull request by slavexx**

### CODE REVIEW

**Comment by slavexx** [comment]
@Orange-hanter we have unresolved moment with 'IRepository' class. Check in issue #574

**Comment by slavexx** [comment]
I pushed penultimate commit. Last commit aimed at removing intersections of lines. After that, i want to recheck the entire diagram, then PR will be opened

**Comment by slavexx** [comment]
There is problem with displaying the "composition" relation (it displayed as simple line). Fix will be.

File: DataAccess/Public/Include/DataAccess/SQLStatements.hpp [hunk 1]
```diff
@@ -17,9 +17,9 @@ class QueryBuilder;
  * @class SQLWhereCondition
  * @brief SQLWhereCondition class.
  * @details This class is a secondary. It needs for \
- *    SQLSelect, SQLUpdate and SQLUpdate. \
```
Comment (memorisecodead) [inline DataAccess/Public/Include/DataAccess/SQLStatements.hpp]: Lol, how is it passed review :3

**Review by KliKonNik — APPROVED** [review]

File: DataAccess/Public/Include/DataAccess/AbstractRepositoryContainer.hpp [hunk 2]
```diff
@@ -12,7 +12,10 @@
 namespace DataAccess
 {
 /* @class AbstractRepositoryContainer
- *  @brief An abstract class for server and client global repositories.
+ * @brief An abstract class for server and client global repositories.
+ * @details Consists adapter to chosen database, and map of  registered repositories.
```
Comment (ivanboika) [inline DataAccess/Public/Include/DataAccess/AbstractRepositoryContainer.hpp]: I'm sorry, but consists of

File: DataAccess/Public/Include/DataAccess/IServerRepositories.hpp [hunk 3]
```diff
@@ -34,7 +35,8 @@ struct IMessagesRepository : IMasterRepository
 };
 
 /**
- * @brief The IRepliesRepository abstract subclass from IMasterRepository
+ * @class IRepliesRepository
+ * @brief Implements repies processing
```
Comment (ivanboika) [inline DataAccess/Public/Include/DataAccess/IServerRepositories.hpp]: typo (replies)

**Review by ivanboika — COMMENTED** [review]
Everything is almost done, just few typos

File: DataAccess/Public/Include/DataAccess/IServerRepositories.hpp (hunk 3 above)
Comment (slavexx) [reply DataAccess/Public/Include/DataAccess/IServerRepositories.hpp]: Fixed

File: DataAccess/Public/Include/DataAccess/AbstractRepositoryContainer.hpp (hunk 2 above)
Comment (slavexx) [reply DataAccess/Public/Include/DataAccess/AbstractRepositoryContainer.hpp]: Fixed

**Review by ivanboika — APPROVED** [review]
Well done

**Review by memorisecodead — APPROVED** [review]

File: DataAccess/Public/Include/DataAccess/IMasterRepository.hpp [hunk 4]
```diff
@@ -7,7 +7,9 @@ namespace DataAccess
 {
 
 /**
+*  @class IMasterRepository
  * @brief The IMasterRepository base abstract class
+ * @details Class used as a base for client and server sides
```
Comment (Orange-hanter) [inline DataAccess/Public/Include/DataAccess/IMasterRepository.hpp]: Base for what?

File: DataAccess/Public/Include/DataAccess/IServerRepositories.hpp [hunk 5]
```diff
@@ -11,7 +11,8 @@
 namespace DataAccess
 {
 /**
-* @brief The IMessagesRepository abstract subclass from IMasterRepository.
+* @class IMessagesRepository
+* @brief Implements message processing
```
Comment (Orange-hanter) [inline DataAccess/Public/Include/DataAccess/IServerRepositories.hpp]: What sort of processing do you mean?

File: DataAccess/Public/Include/DataAccess/IMasterRepository.hpp (hunk 4 above)
Comment (Orange-hanter) [reply DataAccess/Public/Include/DataAccess/IMasterRepository.hpp]: Also, where we should use that abstract class?

File: DataAccess/Public/Include/DataAccess/IClientRepositories.hpp [hunk 6]
```diff
@@ -6,7 +6,12 @@
 namespace DataAccess
 {
 struct MessageInfo;
-
+    /**
+    * @class IMessageRepository
+    * @brief Client message repository
+    * @details Contains methods for client side for
```
Comment (Orange-hanter) [inline DataAccess/Public/Include/DataAccess/IClientRepositories.hpp]: All right, we locate on which side this interface should be used. But still we haven't any idea what for and how we should implement this class.

File: DataAccess/Public/Include/DataAccess/IAdapter.hpp [hunk 7]
```diff
@@ -7,6 +7,12 @@
 
 namespace DataAccess
 {
+    /**
+    * @class IAdapter
+    * @brief An abstract class manages the connection to database
+    * @details Class manages the connection to database, stores options,
```
Comment (Orange-hanter) [inline DataAccess/Public/Include/DataAccess/IAdapter.hpp]: More details are needed. What mean "manages", it asks for some examples. Also, missed information about _query()_ method.
Where this adapter should be used are also unknown.

File: DataAccess/Public/Include/DataAccess/AbstractRepositoryContainer.hpp [hunk 8]
```diff
@@ -12,8 +12,11 @@
 
 namespace DataAccess
 {
-/*  @class AbstractRepositoryContainer
- *  @brief An abstract class for server and client global repositories.
+/* @class AbstractRepositoryContainer
+ * @brief An abstract class for server and client global repositories.
+ * @details Consists of adapter to chosen database, and map of  registered repositories.
+ *  This class used for registration and getting registered repositories,
+ *  determined by further inheritance
```
Comment (Orange-hanter) [inline DataAccess/Public/Include/DataAccess/AbstractRepositoryContainer.hpp]: 1. Remove focus from client and server. What functionality does these class provide ?
2. Explain which class-repository could be registered

File: DataAccess/Public/Include/DataAccess/QueryBuilder.hpp [hunk 9]
```diff
@@ -14,7 +14,9 @@ namespace DataAccess
 {
 /**
  * @class QueryBuilder
- * @brief QueryBuilder class.
+ * @brief Generates SQL query.
```
Comment (Orange-hanter) [inline DataAccess/Public/Include/DataAccess/QueryBuilder.hpp]: "Generate SQL query" is too wide an explanation. It is more of a helper class that provides an interface for building SQL queries.
A few examples can be given here.

File: DataAccess/Public/Include/DataAccess/SQLStatements.hpp [hunk 10]
```diff
@@ -18,7 +18,7 @@ class QueryBuilder;
 * @class SQLWhereCondition
 * @brief SQLWhereCondition class.
 * @details This class is a secondary. It needs for \
-* SQLSelect, SQLUpdate and SQLUpdate. \
+* SQLSelect, SQLUpdate and SQLDelete. \
```
Comment (Orange-hanter) [inline DataAccess/Public/Include/DataAccess/SQLStatements.hpp]: Clear dashes could have sense.

**Review by Orange-hanter — CHANGES_REQUESTED** [review]
In summary, documentation isn't great. Try to explain more details. Remember, you will forget half of that in weak. And another developer didn't curse you at night.

File: DataAccess/Public/Include/DataAccess/IAdapter.hpp (hunk 7 above)
Comment (slavexx) [reply DataAccess/Public/Include/DataAccess/IAdapter.hpp]: We have information about *query()* in PostgreAdapter.hpp. Do i need to copy it here?

Comment (Orange-hanter) [reply DataAccess/Public/Include/DataAccess/IAdapter.hpp]: You could relate to other part of documentation.

File: DataAccess/Public/Include/DataAccess/IMasterRepository.hpp (hunk 4 above)
Comment (slavexx) [reply DataAccess/Public/Include/DataAccess/IMasterRepository.hpp]: Done

File: DataAccess/Public/Include/DataAccess/IServerRepositories.hpp (hunk 5 above)
Comment (slavexx) [reply DataAccess/Public/Include/DataAccess/IServerRepositories.hpp]: Done

File: DataAccess/Public/Include/DataAccess/IClientRepositories.hpp (hunk 6 above)
Comment (slavexx) [reply DataAccess/Public/Include/DataAccess/IClientRepositories.hpp]: Done

File: DataAccess/Public/Include/DataAccess/IAdapter.hpp (hunk 7 above)
Comment (slavexx) [reply DataAccess/Public/Include/DataAccess/IAdapter.hpp]: Done

File: DataAccess/Public/Include/DataAccess/AbstractRepositoryContainer.hpp (hunk 8 above)
Comment (slavexx) [reply DataAccess/Public/Include/DataAccess/AbstractRepositoryContainer.hpp]: Done

File: DataAccess/Public/Include/DataAccess/QueryBuilder.hpp (hunk 9 above)
Comment (slavexx) [reply DataAccess/Public/Include/DataAccess/QueryBuilder.hpp]: Done

File: DataAccess/Public/Include/DataAccess/SQLStatements.hpp (hunk 10 above)
Comment (slavexx) [reply DataAccess/Public/Include/DataAccess/SQLStatements.hpp]: Done

File: DataAccess/Public/Include/DataAccess/IAdapter.hpp [hunk 11]
```diff
@@ -24,6 +31,10 @@ class IAdapter
     template <typename T>
     static std::shared_ptr<T> getInstance(const std::string_view& options = {}) { return T::Instance(options); }
 
+    /**
+    * @brief Method for executing SQL queries.
+    * @details For more details, see PostgreAdapter.hpp
+    */
```
Comment (slavexx) [inline DataAccess/Public/Include/DataAccess/IAdapter.hpp]: more info

File: DataAccess/Public/Include/DataAccess/IMasterRepository.hpp [hunk 12]
```diff
@@ -7,7 +7,10 @@ namespace DataAccess
 {
 
 /**
+*  @class IMasterRepository
  * @brief The IMasterRepository base abstract class
+ * @details Class used as a base of interfaces for repositories on client
+ * and server sides.
```
Comment (slavexx) [inline DataAccess/Public/Include/DataAccess/IMasterRepository.hpp]: for what?

File: DataAccess/Public/Include/DataAccess/IServerRepositories.hpp [hunk 13]
```diff
@@ -11,7 +11,10 @@
 namespace DataAccess
 {
 /**
-* @brief The IMessagesRepository abstract subclass from IMasterRepository.
+* @class IMessagesRepository
+* @brief Implements message processing on server side.
+* @Class include methods for working with messages from database
```
Comment (slavexx) [inline DataAccess/Public/Include/DataAccess/IServerRepositories.hpp]: Class -> detailed

Comment (memorisecodead) [reply DataAccess/Public/Include/DataAccess/IServerRepositories.hpp]: looks good

**Comment by slavexx** [comment]
IMessageRepository (*IClientRepositoies.hpp*) isn`t used in the implementation for PostgreSQL, but used for SQLite DB, which is base on Qt

File: DataAccess/Public/Include/DataAccess/AbstractRepositoryManager.hpp [hunk 14]
```diff
@@ -58,12 +56,12 @@ namespace DataAccess
...
     public:
         /**
-         * @brief   Exists for add methods in the queue.
-         *		    Pass all arguments (args) through 'fmt()' function!!!
-         * @param   (Template arg) priority - priority of request.
-         * @param   MethodReference - just a pointer to member function needed to add to the queue.
-         * @param   args - list of args that passing to member function.
-         * @return  Return future object from repository member function.
+         * @brief Exists for add methods in the queue.
+         *	@warning Pass all arguments (args) through 'fmt()' function
```
Comment (slavexx) [inline DataAccess/Public/Include/DataAccess/AbstractRepositoryManager.hpp]: fix tabulation

**Comment by memorisecodead** [comment]
I think we should raise the issue of rebuilding everything to a different pattern

**Review by memorisecodead — COMMENTED** [review]
I think we should raise the issue of rebuilding everything to a different pattern

**Comment by f1nal3** [comment]
Hey, @slavexx , since we're on the verge of a new release, please answer the following questions:
1) Do you still work on this PR?
2) How long will it take you to finish it?
3) Can it be reviewed and merged now?
4) How much will we lose if the just discard it?

**Comment by slavexx** [comment]
@f1nal3,
    I'm waiting review from Daniil on written documentation. At the moment, work on PR is finished (unless changes are required in future). I`m not sure, that it's a good idea to merge possible incorrect documentation. As far as I know, in future we will implement ORM and this will incur changes in structure of database access. In this case, by skipping this PR we will not lose anything.

**Review by Orange-hanter — APPROVED** [review]

**Pull request merged**